From cc616d20cd7e6346feb4fa6682ba9923911f77c3 Mon Sep 17 00:00:00 2001
From: Govindaraj Saminathan <quic_gsaminat@quicinc.com>
Date: Fri, 29 Nov 2024 14:03:01 +0530
Subject: [PATCH] QSDK:HOSTAPD:CP: Fix crash in MLO by adjusting ucode shutdown
 sequence

A crash occurs when accessing `hostapd_ucode_free_bss()` for the
second link in an MLO scenario. This happens because `ucode_bss_remove`
is accessed after `ucode_shutdown`, leading to improper ucode access.

Move `hostapd_ucode_free()` sequence to the end, after all ucode
BSS instances (`bss_remove`) have been removed, to prevent crashes.

Patch-dependency: 600-ubus_support.patch
patch-work: None
Feature: UCODE

Signed-off-by: Govindaraj Saminathan <quic_gsaminat@quicinc.com>
---
 hostapd/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hostapd/main.c b/hostapd/main.c
index 6870f2e..f700d5e 100644
--- a/hostapd/main.c
+++ b/hostapd/main.c
@@ -1066,7 +1066,6 @@ int main(int argc, char *argv[])
 	ret = 0;
 
  out:
-	hostapd_ucode_free();
 	hostapd_global_ctrl_iface_deinit(&interfaces);
 	/* Sending deauth to all stations before deinit */
 	hostapd_deauthenticate_stations(&interfaces);
@@ -1085,6 +1084,7 @@ int main(int argc, char *argv[])
 	interfaces.count = 0;
 
 	hostapd_global_cleanup_mld(&interfaces);
+	hostapd_ucode_free();
 
 #ifdef CONFIG_DPP
 	dpp_global_deinit(interfaces.dpp);
-- 
2.34.1

